test/e2e/uniswapTrade.test.ts

```diff
@@ -0,0 +1,230 @@
+import ERC20 from "@openzeppelin/contracts/build/contracts/ERC20PresetMinterPauser.json";
+import UniswapV2Factory from "@uniswap/v2-core/build/UniswapV2Factory.json";
+import UniswapV2Pair from "@uniswap/v2-core/build/UniswapV2Pair.json";
+import { expect } from "chai";
+import { Contract, TypedDataDomain, Wallet } from "ethers";
+import { ethers, waffle } from "hardhat";
+
+import {
+  OrderKind,
+  SettlementEncoder,
+  SigningScheme,
+  domain,
+} from "../../src/ts";
+
+import { deployTestContracts } from "./fixture";
+
+describe("E2E: Should Trade Surplus With Uniswap", () => {
+  let deployer: Wallet;
+  let solver: Wallet;
+  let pooler: Wallet;
+  let traders: Wallet[];
+
+  let settlement: Contract;
+  let allowanceManager: Contract;
+  let domainSeparator: TypedDataDomain;
+
+  let weth: Contract;
+  let usdt: Contract;
+  let uniswapPair: Contract;
+  let isWethToken0: boolean;
+
+  beforeEach(async () => {
+    const deployment = await deployTestContracts();
+
+    ({
+      deployer,
+      settlement,
+      allowanceManager,
+      wallets: [solver, pooler, ...traders],
+    } = deployment);
+
+    const { authenticator, owner } = deployment;
+    await authenticator.connect(owner).addSolver(solver.address);
+
+    const { chainId } = await ethers.provider.getNetwork();
+    domainSeparator = domain(chainId, settlement.address);
+
+    weth = await waffle.deployContract(deployer, ERC20, ["WETH", 18]);
+    usdt = await waffle.deployContract(deployer, ERC20, ["USDT", 6]);
+
+    const uniswapFactory = await waffle.deployContract(
+      deployer,
+      UniswapV2Factory,
+      [deployer.address],
+    );
+    await uniswapFactory.createPair(weth.address, usdt.address);
+    uniswapPair = new Contract(
+      await uniswapFactory.getPair(weth.address, usdt.address),
+      UniswapV2Pair.abi,
+      deployer,
+    );
+
```

Comment:
Somewhere, possibly here, there should be another `settle` transaction just to make the two one-time approvals needed to have the settlement contract trade with Uniswap.

Reply:
We actually do `transfer` to the Uniswap pair and then a `swap` directly. I think we could also set up a end 2 end test where we use the same transaction as a user would on the uniswap app (with a limit price and a route for example). In that case we would need to approval that you suggested.

Reply:
Ah, good point!